sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -108,11 +108,15 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
   {
     std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
 
-    Command *NewCmd = static_cast<Command *>(NewEvent->getCommand());
+    // enqueueCommand() func below may throw an exception, so use shared_ptr
+    // here to avoid memory leak
+    std::shared_ptr<Command> NewCmd(
```

Comment:
The smart ptr is owner of the memory of command now. It'll free the memory upon destructor call.
Deallocation is only performed in some corner case here.
Hence, a call to `release()` (which isn't present for `shared_ptr`) method is required in the end in general case and call to `reset()` in the corner case.

Reply:
BTW, corner case is the case when exception is thrown from `GraphProcessor::enqueueCommand()` function, and stack starts to unwind. I think there is no place to add a call to `reset()` function in such case. What do you think?

Reply:
The corner case I was talking about is the `delete NewCmd`, which was removed: https://github.com/intel/llvm/pull/3749/files#diff-583d91db0b0dd79af9e56cdc4550fe76ee97271aa73955241b7a316907f63c5eL128

As for the exception thrown by `GraphProcessor::enqueueCommand()`, is it an option to catch the exception?

Reply:
Fixed.

> As for the exception thrown by `GraphProcessor::enqueueCommand()`, is it an option to catch the exception?

Such async exception should be caught on the user's side using async exception handler passed to `queue`.